File: code/05-dynamic-programming/c/min_path_sum.c
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct { long long sum; size_t *rows, *cols, length; } PathResult;

/* Top-down with memoization (sentinel `seen[i][j]`), matching the
 * MinPathMemo pseudocode's boundary recurrence exactly: base cell (0,0),
 * first row/column accumulate, interior cell is M[i][j]+min(up,left). New
 * for this book -- the naive/memoized recursive forms don't exist in the
 * canonical lecture-notes C, only the bottom-up loop version does. */

// snippet:min-path-memo:start
static long long min_path_memo_rec(const long long *a, size_t cols, int i, int j, long long *memo, bool *seen) {
    size_t k = (size_t)i * cols + (size_t)j;
    if (seen[k]) return memo[k];
    long long result;
    if (i == 0 && j == 0) {
        result = a[k];
    } else if (i == 0) {
        result = min_path_memo_rec(a, cols, i, j - 1, memo, seen) + a[k];
    } else if (j == 0) {
        result = min_path_memo_rec(a, cols, i - 1, j, memo, seen) + a[k];
    } else {
        long long up = min_path_memo_rec(a, cols, i - 1, j, memo, seen);
        long long left = min_path_memo_rec(a, cols, i, j - 1, memo, seen);
        result = (up <= left ? up : left) + a[k];
    }
    seen[k] = true;
    memo[k] = result;
    return result;
}

static bool min_path_memo(const long long *a, size_t rows, size_t cols, long long *out) {
    if (!a || !out || rows == 0 || cols == 0) return false;
    size_t n = rows * cols;
    long long *memo = malloc(n * sizeof(*memo));
    bool *seen = calloc(n, sizeof(*seen));
    if (!memo || !seen) { free(memo); free(seen); return false; }
    *out = min_path_memo_rec(a, cols, (int)rows - 1, (int)cols - 1, memo, seen);
    free(memo); free(seen);
    return true;
}
// snippet:min-path-memo:end

/* Bottom-up, row-major evaluation order (Bottom-Up Evaluation Order
 * pseudocode). Reused verbatim from lecture-notes/code/lecture05/c/min_path_sum.c
 * (min_path), renamed to match this chapter's pseudocode block name. */

// snippet:matrix-bottom-up:start
static bool min_path_bottom_up(const long long *a, size_t rows, size_t cols, PathResult *out) {
    if (!a || !out || rows == 0 || cols == 0) return false;
    size_t n = rows * cols;
    long long *dp = malloc(n * sizeof(*dp));
    unsigned char *parent = calloc(n, 1); /* 1=up, 2=left */
    if (!dp || !parent) { free(dp); free(parent); return false; }
    dp[0] = a[0];
    for (size_t j=1;j<cols;++j){dp[j]=dp[j-1]+a[j];parent[j]=2;}
    for (size_t i=1;i<rows;++i){dp[i*cols]=dp[(i-1)*cols]+a[i*cols];parent[i*cols]=1;}
    for (size_t i=1;i<rows;++i) for(size_t j=1;j<cols;++j){
        size_t k=i*cols+j; long long up=dp[k-cols], left=dp[k-1];
        if (up <= left) { dp[k]=up+a[k]; parent[k]=1; } /* tie: up */
        else { dp[k]=left+a[k]; parent[k]=2; }
    }
    size_t len=rows+cols-1;
    out->rows=malloc(len*sizeof(*out->rows)); out->cols=malloc(len*sizeof(*out->cols));
    if(!out->rows||!out->cols){free(out->rows);free(out->cols);free(dp);free(parent);return false;}
    size_t i=rows-1,j=cols-1,pos=len;
    while(pos){--pos;out->rows[pos]=i;out->cols[pos]=j;if(i==0&&j==0)break;if(parent[i*cols+j]==1)--i;else --j;}
    out->sum=dp[n-1];out->length=len;free(dp);free(parent);return true;
}
// snippet:matrix-bottom-up:end

static void release(PathResult *r) { free(r->rows); free(r->cols); }

int main(void) {
    const long long m[] = {6,7,12,5, 5,3,11,18, 7,17,3,3, 8,10,14,9};
    long long memo_sum = -1;
    min_path_memo(m, 4, 4, &memo_sum);
    printf("memo sum: %lld\n", memo_sum);

    PathResult r = {0};
    min_path_bottom_up(m, 4, 4, &r);
    printf("bottom_up sum: %lld\n", r.sum);
    printf("path:");
    for (size_t k = 0; k < r.length; ++k) printf(" (%zu,%zu)", r.rows[k] + 1, r.cols[k] + 1);
    printf("\n");
    release(&r);
    return 0;
}

File: code/05-dynamic-programming/java/MinPathSum.java
public final class MinPathSum {
    public record Result(long sum, int[][] path) {}
    private MinPathSum() {}

    // Top-down with memoization (sentinel `seen[i][j]`), matching the
    // MinPathMemo pseudocode's boundary recurrence exactly. New for this
    // book -- the canonical Java only has the bottom-up loop version.
    // snippet:min-path-memo:start
    private static long minPathMemoRec(long[][] a, int i, int j, long[][] memo, boolean[][] seen) {
        if (seen[i][j]) return memo[i][j];
        long result;
        if (i == 0 && j == 0) {
            result = a[0][0];
        } else if (i == 0) {
            result = minPathMemoRec(a, i, j - 1, memo, seen) + a[i][j];
        } else if (j == 0) {
            result = minPathMemoRec(a, i - 1, j, memo, seen) + a[i][j];
        } else {
            long up = minPathMemoRec(a, i - 1, j, memo, seen);
            long left = minPathMemoRec(a, i, j - 1, memo, seen);
            result = (up <= left ? up : left) + a[i][j];
        }
        seen[i][j] = true;
        memo[i][j] = result;
        return result;
    }

    public static long minPathMemo(long[][] a) {
        int r = a.length, c = a[0].length;
        long[][] memo = new long[r][c];
        boolean[][] seen = new boolean[r][c];
        return minPathMemoRec(a, r - 1, c - 1, memo, seen);
    }
    // snippet:min-path-memo:end

    // Bottom-up, row-major evaluation order. Reused verbatim from
    // lecture-notes/code/lecture05/java/MinPathSum.java (solve), renamed to
    // match this chapter's pseudocode block name. Rectangular matrix,
    // right/down. Theta(rows*cols) time/table space.
    // snippet:matrix-bottom-up:start
    public static Result minPathBottomUp(long[][] a) {
        if (a == null || a.length == 0 || a[0] == null || a[0].length == 0) {
            throw new IllegalArgumentException();
        }
        int rows = a.length, cols = a[0].length;
        for (long[] row : a) {
            if (row == null || row.length != cols) throw new IllegalArgumentException();
        }

        long[][] dp = new long[rows][cols];
        byte[][] parent = new byte[rows][cols]; // 1=up, 2=left
        dp[0][0] = a[0][0];
        for (int j = 1; j < cols; j++) {
            dp[0][j] = Math.addExact(dp[0][j - 1], a[0][j]);
            parent[0][j] = 2;
        }
        for (int i = 1; i < rows; i++) {
            dp[i][0] = Math.addExact(dp[i - 1][0], a[i][0]);
            parent[i][0] = 1;
        }
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                long up = dp[i - 1][j], left = dp[i][j - 1];
                if (up <= left) {
                    dp[i][j] = Math.addExact(up, a[i][j]);
                    parent[i][j] = 1;
                } else {
                    dp[i][j] = Math.addExact(left, a[i][j]);
                    parent[i][j] = 2;
                }
            }
        }

        int[][] path = new int[rows + cols - 1][2];
        int i = rows - 1, j = cols - 1;
        for (int k = path.length - 1; k >= 0; k--) {
            path[k][0] = i;
            path[k][1] = j;
            if (i == 0 && j == 0) break;
            if (parent[i][j] == 1) i--; else j--;
        }
        return new Result(dp[rows - 1][cols - 1], path);
    }
    // snippet:matrix-bottom-up:end

    public static void main(String[] args) {
        long[][] m = {{6,7,12,5},{5,3,11,18},{7,17,3,3},{8,10,14,9}};
        System.out.println("memo sum: " + minPathMemo(m));
        Result r = minPathBottomUp(m);
        System.out.println("bottom_up sum: " + r.sum());
        StringBuilder sb = new StringBuilder("path:");
        for (int[] cell : r.path()) sb.append(" (").append(cell[0] + 1).append(",").append(cell[1] + 1).append(")");
        System.out.println(sb);
    }
}
